source/MaterialXShaderGen/Shader.cpp

```diff
@@ -12,12 +12,18 @@
 namespace MaterialX
 {
 
+const string Shader::GLOBAL_SCOPE = "Globals";
+const string Shader::SHADER_INTERFACE = "ShaderInterface";
+
 Shader::Shader(const string& name)
     : _name(name)
     , _rootGraph(nullptr)
     , _activeStage(PIXEL_STAGE)
+    , _appData("AppData", "ad")
 {
-    _stages.resize(numStages());
+    // Create default uniform blocks
+    createUniformBlock(GLOBAL_SCOPE);
```

Comment:
Yeah, this is just a way to separate "internal" uniforms needed by node implementations from "public" uniforms created from the graph/shader interface. For example the shader interface uniforms should be visible in a UI, and they should only be needed in the pixel shader stage. 

I added an instance name "si" for the ShaderInterface block since my idea was to declare this as a uniform block in the pixel shader:

'uniform ShaderInterface { ... } si;'

but then I discovered that textures/samplers can't be declared inside a uniform block. So in code they are still all declared at global scope. So yeah it's mainly a way to be able to access them separately.

But I think you are right that this could be improved on for clarity.